packages/compile-solidity/compileWithPragmaAnalysis.js

```diff
@@ -0,0 +1,139 @@
+const CompilerSupplier = require("./compilerSupplier");
+const Config = require("@truffle/config");
+const semver = require("semver");
+const Profiler = require("./profiler");
+const fse = require("fs-extra");
+const { run } = require("./run");
+const OS = require("os");
+const cloneDeep = require("lodash.clonedeep");
+
+const getSemverExpression = source => {
+  return source.match(/pragma solidity(.*);/)[1].trim();
+};
+
+const getSemverExpressions = sources => {
+  return sources.map(source => getSemverExpression(source));
+};
+
+const validateSemverExpressions = semverExpressions => {
+  const { validRange } = semver;
+  for (const expression of semverExpressions) {
+    if (semver.validRange(expression) === null) {
+      const message = `Invalid semver expression (${expression}) found in` +
+        `one of your contract's imports.`;
+      throw new Error(message);
+    }
+  }
+};
+
+// takes an array of versions and an array of semver expressions
+// returns a version of the compiler or undefined if none can be found
+const findNewestSatisfyingVersion = ({ solcReleases, semverExpressions }) => {
```

Comment:
Haha, you didn't end up using my complicated intersection function, I see XD

Also: If you're using an arrow function, but the function body is just a single return statement, no need to bother with braces and `return`, right? :)

Also: I think this function can be simplified a little by use of the semver function `maxSatisfying`:

```
const findNewestSatisfyingVersion = ({ solcReleases, semverExpressions }) =>
  maxSatisfying(
    semverExpressions.map(range => maxSatisfying(solcReleases, range)),
    ""
  )
```

The inner `maxSatisfying` does what it says it does -- finds the maximum solc version among the given releases that satisfies the given range.  The outer one is maybe a bit of a hack... I'm just using it to take the maximum of the results, passing `""` as the range to ensure they all satisfy it. :)

Reply:
I don't think this is what we want though. We need to find the maximum version that satisfies every range. If you take the max for every range and then pick the highest then it is likely that you'll get a range that only satisfies a subset of the ranges input and therefore won't be compatible with some range.

For example, say you have `["0.6.0", ">0.5.0 <=0.7.2"]`. Won't this method return "0.7.2" which won't satisfy `"0.6.0"`?

Reply:
Oh, sily me, you're absolutely right.  Blargh.  Don't know why I keep getting this mixed up.

Yeah, nevermind, you have it right.  We could also use my ridiculous intersection function, but, uh, this way is probably better.